File: JavaScript/uri-1034.js
// URI - 1034: https://www.urionlinejudge.com.br/judge/pt/problems/view/1034

const input = require("fs").readFileSync("input.txt", "utf-8");
const lines = input.split("\n");

let numberOfInstances = lines.shift();
let blocksInformation, desiredLength, blocks, minimumNumberOfBlocks;

for (let i = 0; i < numberOfInstances; i++) {
  blocksInformation = lines.shift().split(" ");
  desiredLength = parseInt(blocksInformation[1]);
  blocks = lines.shift().split(" ");
  blocks = blocks.map(block => parseInt(block));
  minimumNumberOfBlocks = searchTheMinimumNumberOfBlocks(blocks, desiredLength);
  console.log(minimumNumberOfBlocks);
}

function searchTheMinimumNumberOfBlocks(blocks, desiredLength) {
  requiredBlocks = [];
  requiredBlocks[0] = 0;
  if (requiredBlocks.length >= desiredLength)
    return requiredBlocks[desiredLength];
  else {
    for (let i = 1; i <= desiredLength; i++) {
      requiredBlocks[i] = 1000000;
      for (let j = 0; j < blocks.length; j++) {
        if (i - blocks[j] >= 0)
          requiredBlocks[i] = Math.min(requiredBlocks[i], requiredBlocks[i - blocks[j]] + 1);
      }
    }
  }
  return requiredBlocks[desiredLength];
}

File: Ruby/uri-1034.rb
# frozen_string_literal: true

# URI - 1034: https://www.urionlinejudge.com.br/judge/pt/problems/view/1034

def calculate_the_minimum_number_of_blocks(blocks, desidered_size)
  minimum_number_of_blocks = [0]
  (1..desidered_size).each do |current_size|
    minimum_number_of_blocks[current_size] = INFINITE
    select_blocks(blocks, minimum_number_of_blocks, current_size)
  end
  minimum_number_of_blocks[desidered_size]
end

def select_blocks(blocks, minimum_number_of_blocks, current_size)
  (0...blocks.length).each do |block_index|
    next unless current_size - blocks[block_index] >= 0

    minimum_number_of_blocks[current_size] = [
      minimum_number_of_blocks[current_size],
      minimum_number_of_blocks[current_size - blocks[block_index]] + 1
    ].min { |a, b| a <=> b }
  end
end

INFINITE = 99_999
number_of_instances = gets.to_i
number_of_instances.times do
  input = gets.chomp.split(' ')
  blocks = gets.chomp.split(' ').map(&:to_i)
  puts calculate_the_minimum_number_of_blocks(blocks, input[1].to_i)
end
